{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# packages\n",
    "import numpy as np\n",
    "import itertools\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import argparse\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_size = 5E5\n",
    "dengueGenome=\"A\"*100\n",
    "seq_length = len(dengueGenome)\n",
    "alphabet = ['A', 'T', 'G', 'C']\n",
    "mutation_rate = 0.0075 # per gen per individual per site\n",
    "##include mutation rate models\n",
    "fitness_effect_ben = 1.15 # fitness effect if a functional mutation occurs\n",
    "fitness_chance_ben = 0.05 # chance that a mutation has a fitness effect\n",
    "\n",
    "fitness_chance_del = 0.4 # chance that a mutation has a fitness effect\n",
    "fitness_effect_del = 0.5\n",
    "\n",
    "fitness_chance_lethal = 0.03 # chance that a mutation has a fitness effect\n",
    "fitness_effect_lethal = 0 # chance that a mutation has a fitness effect"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# population\n",
    "base_haplotype = dengueGenome\n",
    "pop = {}\n",
    "pop[base_haplotype] = pop_size\n",
    "fitness = {}\n",
    "fitness[base_haplotype] = 1.0\n",
    "history = []\n",
    "\n",
    "# mutation\n",
    "def get_mutation_count():\n",
    "    mean = mutation_rate * pop_size * seq_length\n",
    "    return np.random.poisson(mean)\n",
    "\n",
    "def get_random_haplotype():\n",
    "    haplotypes = list(pop.keys())\n",
    "    frequencies = [x/float(pop_size) for x in pop.values()]\n",
    "    total = sum(frequencies)\n",
    "    frequencies = [x / total for x in frequencies]\n",
    "    return np.random.choice(haplotypes, p=frequencies)\n",
    "\n",
    "def get_mutant(haplotype):\n",
    "    site = np.random.randint(seq_length)\n",
    "    possible_mutations = list(alphabet)\n",
    "    possible_mutations.remove(haplotype[site])\n",
    "    mutation = np.random.choice(possible_mutations)\n",
    "    new_haplotype = haplotype[:site] + mutation + haplotype[site+1:]\n",
    "    return new_haplotype\n",
    "\n",
    "def get_fitness(haplotype):\n",
    "    old_fitness = fitness[haplotype]\n",
    "    if (np.random.random() < fitness_chance_ben):\n",
    "        return old_fitness * fitness_effect_ben\n",
    "    elif(np.random.random() < fitness_chance_lethal):\n",
    "        return old_fitness * fitness_effect_lethal\n",
    "    elif(np.random.random() < fitness_chance_del):\n",
    "        return old_fitness * fitness_effect_del\n",
    "    else:\n",
    "        return old_fitness\n",
    "\n",
    "def mutation_event():\n",
    "    haplotype = get_random_haplotype()\n",
    "    if pop[haplotype] > 1:\n",
    "        pop[haplotype] -= 1\n",
    "        new_haplotype = get_mutant(haplotype)\n",
    "        if new_haplotype in pop:\n",
    "            pop[new_haplotype] += 1\n",
    "        else:\n",
    "            pop[new_haplotype] = 1\n",
    "        if new_haplotype not in fitness:\n",
    "            fitness[new_haplotype] = get_fitness(haplotype)\n",
    "\n",
    "def mutation_step():\n",
    "    mutation_count = get_mutation_count()\n",
    "    for i in range(mutation_count):\n",
    "        mutation_event()\n",
    "\n",
    "# genetic drift and selection\n",
    "def get_offspring_counts():\n",
    "    haplotypes = list(pop.keys())\n",
    "    frequencies = [pop[haplotype]/float(pop_size) for haplotype in haplotypes]\n",
    "    fitnesses = [fitness[haplotype] for haplotype in haplotypes]\n",
    "    weights = [x * y for x,y in zip(frequencies, fitnesses)]\n",
    "    total = sum(weights)\n",
    "    weights = [x / total for x in weights]\n",
    "    return list(np.random.multinomial(pop_size, weights))\n",
    "\n",
    "def offspring_step():\n",
    "    counts = get_offspring_counts()\n",
    "    for (haplotype, count) in zip(list(pop.keys()), counts):\n",
    "        if (count > 0):\n",
    "            pop[haplotype] = count\n",
    "        else:\n",
    "            del pop[haplotype]\n",
    "\n",
    "# simulate\n",
    "def time_step():\n",
    "    mutation_step()\n",
    "    offspring_step()\n",
    "    \n",
    "def simulate():\n",
    "    clone_pop = dict(pop)\n",
    "    history.append(clone_pop)\n",
    "    for i in range(generations):\n",
    "        print (i)\n",
    "        time_step()\n",
    "        clone_pop = dict(pop)\n",
    "        history.append(clone_pop)\n",
    "\n",
    "# plot diversity\n",
    "def get_distance(seq_a, seq_b):\n",
    "    diffs = 0\n",
    "    length = len(seq_a)\n",
    "    assert len(seq_a) == len(seq_b)\n",
    "    for chr_a, chr_b in zip(seq_a, seq_b):\n",
    "        if chr_a != chr_b:\n",
    "            diffs += 1\n",
    "    return diffs / float(length)\n",
    "\n",
    "def get_diversity(population):\n",
    "    haplotypes = list(population.keys())\n",
    "    haplotype_count = len(haplotypes)\n",
    "    diversity = 0\n",
    "    for i in range(haplotype_count):\n",
    "        for j in range(haplotype_count):\n",
    "            haplotype_a = haplotypes[i]\n",
    "            haplotype_b = haplotypes[j]\n",
    "            frequency_a = population[haplotype_a] / float(pop_size)\n",
    "            frequency_b = population[haplotype_b] / float(pop_size)\n",
    "            frequency_pair = frequency_a * frequency_b\n",
    "            diversity += frequency_pair * get_distance(haplotype_a, haplotype_b)\n",
    "    return diversity\n",
    "\n",
    "def get_diversity_trajectory():\n",
    "    trajectory = [get_diversity(generation) for generation in history]\n",
    "    return trajectory\n",
    "\n",
    "def diversity_plot(xlabel=\"generation\"):\n",
    "    mpl.rcParams['font.size']=14\n",
    "    trajectory = get_diversity_trajectory()\n",
    "    plt.plot(trajectory, \"#447CCD\")    \n",
    "    plt.ylabel(\"diversity\")\n",
    "    plt.xlabel(xlabel)\n",
    "    \n",
    "# plot divergence\n",
    "def get_divergence(population):\n",
    "    haplotypes = list(population.keys())\n",
    "    divergence = 0\n",
    "    for haplotype in haplotypes:\n",
    "        frequency = population[haplotype] / float(pop_size)\n",
    "        divergence += frequency * get_distance(base_haplotype, haplotype)\n",
    "    return divergence\n",
    "\n",
    "def get_divergence_trajectory():\n",
    "    trajectory = [get_divergence(generation) for generation in history]\n",
    "    return trajectory\n",
    "\n",
    "def divergence_plot(xlabel=\"generation\"):\n",
    "    mpl.rcParams['font.size']=14\n",
    "    trajectory = get_divergence_trajectory()\n",
    "    plt.plot(trajectory, \"#447CCD\")\n",
    "    plt.ylabel(\"divergence\")\n",
    "    plt.xlabel(xlabel) \n",
    "\n",
    "# plot trajectories\n",
    "def get_frequency(haplotype, generation):\n",
    "    pop_at_generation = history[generation]\n",
    "    if haplotype in pop_at_generation:\n",
    "        return pop_at_generation[haplotype]/float(pop_size)\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    "def get_trajectory(haplotype):\n",
    "    trajectory = [get_frequency(haplotype, gen) for gen in range(generations)]\n",
    "    return trajectory\n",
    "\n",
    "def get_all_haplotypes():\n",
    "    haplotypes = set()   \n",
    "    for generation in history:\n",
    "        for haplotype in generation:\n",
    "            haplotypes.add(haplotype)\n",
    "    return haplotypes\n",
    "\n",
    "def stacked_trajectory_plot(xlabel=\"generation\"):\n",
    "    colors_lighter = [\"#A567AF\", \"#8F69C1\", \"#8474D1\", \"#7F85DB\", \"#7F97DF\", \"#82A8DD\", \"#88B5D5\", \"#8FC0C9\", \"#97C8BC\", \"#A1CDAD\", \"#ACD1A0\", \"#B9D395\", \"#C6D38C\", \"#D3D285\", \"#DECE81\", \"#E8C77D\", \"#EDBB7A\", \"#EEAB77\", \"#ED9773\", \"#EA816F\", \"#E76B6B\"]\n",
    "    mpl.rcParams['font.size']=18\n",
    "    haplotypes = get_all_haplotypes()\n",
    "    trajectories = [get_trajectory(haplotype) for haplotype in haplotypes]\n",
    "    plt.stackplot(range(generations), trajectories, colors=colors_lighter)\n",
    "    plt.ylim(0, 1)\n",
    "    plt.ylabel(\"frequency\")\n",
    "    plt.xlabel(xlabel)\n",
    "\n",
    "# plot snp trajectories\n",
    "def get_snp_frequency(site, generation):\n",
    "    minor_allele_frequency = 0.0\n",
    "    pop_at_generation = history[generation]\n",
    "    for haplotype in list(pop_at_generation.keys()):\n",
    "        allele = haplotype[site]\n",
    "        frequency = pop_at_generation[haplotype] / float(pop_size)\n",
    "        if allele != dengueGenome[site]:\n",
    "            minor_allele_frequency += frequency\n",
    "    return minor_allele_frequency\n",
    "\n",
    "def get_snp_trajectory(site):\n",
    "    trajectory = [get_snp_frequency(site, gen) for gen in range(generations)]\n",
    "    return trajectory\n",
    "\n",
    "def get_all_snps():\n",
    "    snps = set()   \n",
    "    for generation in history:\n",
    "        for haplotype in generation:\n",
    "            for site in range(seq_length):\n",
    "                if haplotype[site] != dengueGenome[site]:\n",
    "                    snps.add(site)\n",
    "    return snps\n",
    "\n",
    "def snp_trajectory_plot(xlabel=\"generation\"):\n",
    "    colors = [\"#781C86\", \"#571EA2\", \"#462EB9\", \"#3F47C9\", \"#3F63CF\", \"#447CCD\", \"#4C90C0\", \"#56A0AE\", \"#63AC9A\", \"#72B485\", \"#83BA70\", \"#96BD60\", \"#AABD52\", \"#BDBB48\", \"#CEB541\", \"#DCAB3C\", \"#E49938\", \"#E68133\", \"#E4632E\", \"#DF4327\", \"#DB2122\"]\n",
    "    mpl.rcParams['font.size']=18\n",
    "    snps = get_all_snps()\n",
    "    trajectories = [get_snp_trajectory(snp) for snp in snps]\n",
    "    data = []\n",
    "    for trajectory, color in zip(trajectories, itertools.cycle(colors)):\n",
    "        data.append(range(generations))\n",
    "        data.append(trajectory)    \n",
    "        data.append(color)\n",
    "    plt.plot(*data)   \n",
    "    plt.ylim(0, 1)\n",
    "    plt.ylabel(\"frequency\")\n",
    "    plt.xlabel(xlabel)\n",
    "\n",
    "# if __name__==\"__main__\":\n",
    "#     parser = argparse.ArgumentParser(description = \"run wright-fisher simulation with mutation and genetic drift\")\n",
    "#     parser.add_argument('--pop_size', type = int, default = 50000, help = \"population size\")\n",
    "#     parser.add_argument('--mutation_rate', type = float, default = 0.000009, help = \"mutation rate\")\n",
    "#     parser.add_argument('--seq_length', type = int, default = len(dengueGenome), help = \"sequence length\")\n",
    "#     parser.add_argument('--generations', type = int, default = 10, help = \"generations\")\n",
    "#     parser.add_argument('--fitness_effect_ben', type = float, default = 1.2, help = \"fitness effect ben\")\n",
    "#     parser.add_argument('--fitness_effect_del', type = float, default = 0.2, help = \"fitness effect del\")\n",
    "#     parser.add_argument('--fitness_chance_ben', type = float, default = 0.01, help = \"fitness chance ben\")    \n",
    "#     parser.add_argument('--fitness_chance_del', type = float, default = 0.1, help = \"fitness chance del\")  \n",
    "#     parser.add_argument('--summary', action = \"store_true\", default = False, help = \"don't plot trajectories\")        \n",
    "\n",
    "#params = parser.parse_args()\n",
    "# pop_size = params.pop_size\n",
    "# mutation_rate = params.mutation_rate\n",
    "# seq_length = params.seq_length\n",
    "# generations = params.generations\n",
    "# fitness_effect_ben = params.fitness_effect_ben\n",
    "# fitness_effect_del = params.fitness_effect_del\n",
    "# fitness_chance_ben = params.fitness_effect_ben\n",
    "# fitness_chance_del = params.fitness_chance_del\n",
    "generations=10\n",
    "\n",
    "simulate()\n",
    "\n",
    "print (\"done\")\n",
    "\n",
    "plt.figure(num=None, figsize=(14, 10), dpi=80, facecolor='w', edgecolor='k')\n",
    "if True:\n",
    "    plt.subplot2grid((2,1), (0,0))\n",
    "    diversity_plot()\n",
    "    plt.subplot2grid((2,1), (1,0))\n",
    "    divergence_plot()\n",
    "else:    \n",
    "    plt.subplot2grid((3,2), (0,0), colspan=2)\n",
    "    stacked_trajectory_plot(xlabel=\"\")\n",
    "    plt.subplot2grid((3,2), (1,0), colspan=2)\n",
    "    snp_trajectory_plot(xlabel=\"\")\n",
    "    plt.subplot2grid((3,2), (2,0))\n",
    "    diversity_plot()\n",
    "    plt.subplot2grid((3,2), (2,1))\n",
    "    divergence_plot()\n",
    "plt.show()        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
